### PR TITLE
Create blur.cpp

### PR DESCRIPTION
**Pull request by Rrytel**

### CODE REVIEW

**Comment by stevenireeves** [comment]
@Rrytel what is the issue in GenMat.cpp?

File: GenMat.cpp [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+        std::cout << A.elements[i * A.width + j];
+        std::cout << " ";
+      }
+      std::cout << std::endl;
+     }
+     std::cout << std::endl;
+     std::cout << std::endl;
+
+
+     dim3 dimBlockTransA(gcd(aT.width,aT.height),gcd(aT.width,aT.height));
```
Comment (stevenireeves) [inline GenMat.cpp]: The block dimension is generally a tunable parameter. I'm not positive using the gcd is really how you want to do this. Also, there's no guarantee here that `dimBlockTransA.x * dimBlockTransA.y <= 1024`

**Comment by stevenireeves** [comment]
For best performance you'll likely want a single kernel doing GEMM. With the computation split among many kernels you can't leverage caching.

**Comment by Rrytel** [comment]
> @Rrytel what is the issue in GenMat.cpp?

I was having issues with the shared memory matrix multiplication giving incorrect values. I tried testing with the naïve kernel and the results are correct but when I make my matrices large I get just zeros

File: GenMat.cpp [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+  dim3 dimBlock(BLOCK_SIZE, BLOCK_SIZE);
+  dim3 dimGrid(B.width / dimBlock.x, A.height / dimBlock.y);
+  // Use HIP Events for timing
+  hipEvent_t start, stop;
+  float time;
+  hipEventCreate(&start);
+  hipEventCreate(&stop);
+  hipEventRecord(start, 0);
+
+  //MatMulKernel<<<dimGrid, dimBlock>>>(dA, dB, dC);
```
Comment (stevenireeves) [inline GenMat.cpp]: Did you mean for this to be commented out?

**Comment by stevenireeves** [comment]
> > @Rrytel what is the issue in GenMat.cpp?
> 
> I was having issues with the shared memory matrix multiplication giving incorrect values. I tried testing with the naïve kernel and the results are correct but when I make my matrices large I get just zeros

In the lds kernel or simple kernel?

File: GenMat.cpp [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+    std::cout << std::endl;
+  }
+   std::cout << std::endl;
+   std::cout << std::endl;
+
+
+  // Multiply A and B
+
+  
+  blockSize = gcd(AbyB.width, AbyB.height);
```
Comment (stevenireeves) [inline GenMat.cpp]: Again, I wouldn't use the gcd, there are instances where the gcd will be 1, and then the shared memory multiplication reduces to the naive kernel.

**Comment by Rrytel** [comment]
> > > @Rrytel what is the issue in GenMat.cpp?
> > 
> > 
> > I was having issues with the shared memory matrix multiplication giving incorrect values. I tried testing with the naïve kernel and the results are correct but when I make my matrices large I get just zeros
> 
> In the lds kernel or simple kernel?

simply kernel

**Comment by Rrytel** [comment]
@stevenireeves I tried the shared mem kernel and the results dont work still. I also tried using your kernel.

**Comment by stevenireeves** [comment]
> @stevenireeves I tried the shared mem kernel and the results dont work still. I also tried using your kernel.

What size? The original kernel wasn't generic, it assumes the size is a multiple of blockSize.

**Comment by Rrytel** [comment]
@stevenireeves 
  int N = 16;
  int M = 32;
  int K = 48;

**Comment by stevenireeves** [comment]
> @stevenireeves int N = 16; int M = 32; int K = 48;

Yeah `N * K` wont be a multiple of `16 * 16`.

**Comment by Rrytel** [comment]
@stevenireeves could I call the kernel on a 48x48 and check if row or col is outside the index we want for a 32x48

**Comment by Rrytel** [comment]
@stevenireeves I still dont get results when N=M=K=16

File: GenMat.cpp [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+ */
+
+int gcd(int a, int b) 
+{
+   if (b == 0)
+   return a;
+   return gcd(b, a % b);
+}
+
+void MatMul(const CMat A, const CMat B, CMat C) {
```
Comment (stevenireeves) [inline GenMat.cpp]: CMat &C

**Review by stevenireeves — CHANGES_REQUESTED** [review]

File: GenMat.cpp (hunk 4 above)
Comment (Rrytel) [reply GenMat.cpp]: i dont use the matmul function. I did get results after I changed the blocksize to 16. I accidentally had it as 1 still

Comment (stevenireeves) [reply GenMat.cpp]: You'll need to pass C by reference here. In the demo it was passed by value which was invoking the copy constructor.

File: GenMat.cpp [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+}
+
+__global__ void GpuCopy(GMat A, GMat C)
+{
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   C.elements[row * C.width + col] =  A.elements[row * A.width + col];
+}
+
+void GEMM(const int alpha, const bool opA, const CMat A, const bool opB, const CMat B, const int beta, const CMat C, CMat D)
```
Comment (stevenireeves) [inline GenMat.cpp]: If you want to retain the values in D you'll still want this by Reference.

Comment (stevenireeves) [inline GenMat.cpp]: If you want to retain the values in D you'll still want this by Reference.

File: GenMat.cpp (hunk 4 above)
Comment (stevenireeves) [reply GenMat.cpp]: Make sure to remove dead code

**Comment by Rrytel** [comment]
@stevenireeves do you want the mat.h and mat.cpp files to be included in the pull

**Pull request merged**